File: algorithms/2_medium/C/04054_count_shadow_pairs_i/stack.py
class Solution:
    def shadowPairs(self, nums: list[int]) -> int:
        stack = []
        cnt = 0

        for num in nums:
            while stack and stack[-1][0] > num:
                stack.pop()

            if stack:
                (top, prev_num_cnt) = stack[-1]
                if top == num:
                    cnt += prev_num_cnt
                    stack.append((num, prev_num_cnt))
                else:
                    cnt += len(stack)
                    stack.append((num, len(stack)))
            else:
                stack.append((num, 0))

        return cnt

File: algorithms/2_medium/C/04054_count_shadow_pairs_i/stack.rs
impl Solution {
    pub fn shadow_pairs(nums: Vec<i32>) -> i64 {
        let mut stack: Vec<(i32, usize)> = Vec::with_capacity(nums.len());
        let mut cnt: usize = 0;

        for num in nums {
            while let Some((top, _)) = stack.last() && *top > num {
                stack.pop();
            }

            if let Some(&(top, prev_num_cnt)) = stack.last() {
                if top == num {
                    cnt += prev_num_cnt;
                    stack.push((num, prev_num_cnt));
                } else {
                    cnt += stack.len();
                    stack.push((num, stack.len()));
                }
            } else {
                stack.push((num, 0));
            }
        }

        cnt.try_into().unwrap()
    }
}
